### PR TITLE
Add FreeRDP

### PR DESCRIPTION
**Pull request by Mno-hime**
Picked current upstream's `HEAD` because even 2.0.0 RCs are too old.

Somewhat related: https://www.illumos.org/issues/11508.

**Testing**
- tested with VRDP from VirtualBox 6.0.8.

### CODE REVIEW

File: components/desktop/freerdp/Makefile [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+	sha256:12374e828745e13f707844c4fc4607094f16513b8b94e6728883690fdb9cdc8e
+COMPONENT_LICENSE=	Apache v2.0
+COMPONENT_LICENSE_FILE=	LICENSE
+
+include $(WS_MAKE_RULES)/prep.mk
+include $(WS_MAKE_RULES)/cmake.mk
+include $(WS_MAKE_RULES)/ips.mk
+
+# https://github.com/FreeRDP/FreeRDP/issues/5504
+CFLAGS += -pthread
```
Comment (pyhalov) [inline components/desktop/freerdp/Makefile]: hm... Can you rise a question in \#illumos, I do not understand this. libpthread is just libc filter...

Comment (pyhalov) [reply components/desktop/freerdp/Makefile]: Perhaps, gcc adds some defines when -pthread is used...

Comment (Mno-hime) [reply components/desktop/freerdp/Makefile]: I think for us it essentially boils down to:
```
CFLAGS += -D_REENTRANT -D_PTHREADS
LDFLAGS += -lpthread -ldl -lc
```

Per https://stackoverflow.com/a/23251828:
```patch
$ gdiff -pruN1 dm.nopthread.txt dm.pthread.txt 
--- dm.nopthread.txt	2019-07-30 10:20:41.275025417 +0000
+++ dm.pthread.txt	2019-07-30 10:20:33.776335769 +0000
@@ -66,2 +66,3 @@
 #define __LDBL_HAS_QUIET_NAN__ 1
+#define _REENTRANT 1
 #define __GNUC__ 6
@@ -144,2 +145,3 @@
 #define __INT_LEAST16_MAX__ 0x7fff
+#define _PTHREADS 1
 #define __DEC64_MANT_DIG__ 16
```

```
$ gcc -dumpspecs | grep '%{pthread'
%{!symbolic:     %{pthreads|pthread:-lpthread}      %{p|pg:-ldl} -lc}
%{pthreads|pthread:-D_REENTRANT -D_PTHREADS}
```

Comment (Mno-hime) [reply components/desktop/freerdp/Makefile]: https://github.com/FreeRDP/FreeRDP/issues/5504#issuecomment-516314097

Can it be (and should it be?) that our OpenSSL is without `-pthreads`?

Comment (pyhalov) [reply components/desktop/freerdp/Makefile]: We have `CONFIGURE_OPTIONS += threads` for openssl... 
Interesting, if it's question of defines vs question of libs. If you compile just with -D_REENTRANT, does it behave sane?

Comment (Mno-hime) [reply components/desktop/freerdp/Makefile]: That alone works fine.

**Comment by Mno-hime** [comment]
There's another slight problem. `xfreerdp` only works with elevated privileges (in GZ), otherwise just shows blank screen for a split second:
```
/2:     open("/dev/timerfd", O_RDWR|O_NONBLOCK)         = 28
/2:     ioctl(28, (('t'<<24)|('f'<<16)|('d'<<8)|1), 0x00000004) Err#1 EPERM [proc_clock_highres]
/2:     close(28)                                       = 0
```

[xfreerdp.truss.pfexec.txt](https://github.com/OpenIndiana/oi-userland/files/3446307/xfreerdp.truss.pfexec.txt)
[xfreerdp.truss.txt](https://github.com/OpenIndiana/oi-userland/files/3446308/xfreerdp.truss.txt)

File: components/desktop/freerdp/Makefile (hunk 1 above)
Comment (pyhalov) [reply components/desktop/freerdp/Makefile]: Can you rise this question at illumos-dev what heritage is there? I mean, why it's needed at all...

Comment (pyhalov) [reply components/desktop/freerdp/Makefile]: And, more funny, defining _REENTRANT, enables posix variants of some functions (like readdir_r()... Does it just have an effect of _XOPEN_SOURCE=600 ?

Comment (Mno-hime) [reply components/desktop/freerdp/Makefile]: That fails early with:
```
FreeRDP-e873e4f0042dd157f68ae99adb4dc9dcd167ca29/winpr/include/winpr/winsock.h:318:48: error: unknown type name 'u_long'
 WINPR_API int _ioctlsocket(SOCKET s, long cmd, u_long* argp);
                                                ^~~~~~
```

Comment (Mno-hime) [reply components/desktop/freerdp/Makefile]: > Can you rise this question at illumos-dev what heritage is there? I mean, why it's needed at all...

I will do later.

Comment (pyhalov) [reply components/desktop/freerdp/Makefile]: Try CPPFLAGS+=$(CPP_XPG6MODE)

Comment (pyhalov) [reply components/desktop/freerdp/Makefile]: It would be good to make it default (at least for oi-userland) one day :)

**Comment by pyhalov** [comment]
Do you mean, freerdp works only as root or only in GZ?

**Comment by Mno-hime** [comment]
Works only as root in GZ. Did not tried outside of GZ.

**Comment by pyhalov** [comment]
This is another question to ask - why high-precision timers are restricted...

**Comment by pyhalov** [comment]
As workaround, make xfreerdp privilege aware, ship corresponding exec_attr and run it with elevated privileges (+clock_highres)

File: components/desktop/freerdp/Makefile (hunk 1 above)
Comment (Mno-hime) [reply components/desktop/freerdp/Makefile]: `CFLAGS += $(XPG6MODE)` does work.

Comment (Mno-hime) [reply components/desktop/freerdp/Makefile]: > `CFLAGS += $(XPG6MODE)` does work.

Should I keep it or is any other variant preferred?

Comment (pyhalov) [reply components/desktop/freerdp/Makefile]: CFLAGS += $(XPG6MODE) is fine.

**Comment by pyhalov** [comment]
What about running as non-privileged user?

**Comment by Mno-hime** [comment]
I though this is it: `Basic Solaris User:solaris:cmd:::/usr/bin/xfreerdp:privs=clock_highres` but  user with default privileges can't even execute it due to `EACCES` (with or without the exec_attr file):

```
$ profiles 
Console User
Suspend To RAM
Suspend To Disk
Brightness
CPU Power Management
Network Autoconf User
Desktop Removable Media User
Basic Solaris User
All

$ pfexec /usr/bin/xfreerdp
pfexec: unable to execute /usr/bin/xfreerdp: Permission denied
```

**Comment by pyhalov** [comment]
pfexec ppriv -e -D /usr/bin/xfreerdp?

**Comment by Mno-hime** [comment]
@pyhalov:
```
{global} newman@lenovo:~ $ cat /etc/security/exec_attr.d/freerdp 
Basic Solaris User:solaris:cmd:RO::/usr/bin/xfreerdp:privs=clock_highres
{global} newman@lenovo:~ $ ppriv -e -D /usr/bin/xfreerdp /size:800x600 /v:localhost:5001
[12:07:54:326] [6263:00000002] [INFO][com.freerdp.client.common.cmdline] - loading channelEx rdpdr
[12:07:54:326] [6263:00000002] [INFO][com.freerdp.client.common.cmdline] - loading channelEx rdpsnd
[12:07:54:326] [6263:00000002] [INFO][com.freerdp.client.common.cmdline] - loading channelEx cliprdr
[12:07:54:502] [6263:00000002] [ERROR][com.freerdp.core.nego] - Error: SSL_CERT_NOT_ON_SERVER
[12:07:54:503] [6263:00000002] [ERROR][com.freerdp.core.nego] - Error: SSL_CERT_NOT_ON_SERVER
[12:07:54:537] [6263:00000002] [INFO][com.freerdp.gdi] - Local framebuffer format  PIXEL_FORMAT_BGRX32
[12:07:54:537] [6263:00000002] [INFO][com.freerdp.gdi] - Remote framebuffer format PIXEL_FORMAT_RGB16
[12:07:54:629] [6263:00000002] [INFO][com.winpr.clipboard] - initialized POSIX local file subsystem
[12:07:54:694] [6263:00000008] [INFO][com.freerdp.channels.rdpsnd.client] - Loaded fake backend for rdpsnd
xfreerdp[6263]: missing privilege "proc_clock_highres" (euid = 101, syscall = 54) needed at secpolicy_clock_highres+0x1d
```

**Comment by Mno-hime** [comment]
@pyhalov any idea bout the `ppriv` output?

**Comment by AndWac** [comment]
From what I read the output of ppriv you will need to add "proc_clock_highres" to the user's privileges. It only has "clock_highres" atm.

**Comment by Mno-hime** [comment]
Right, but still:
```
{global} newman@lenovo:~ $ cat /etc/security/exec_attr.d/freerdp 
Basic Solaris User:solaris:cmd:::/usr/bin/xfreerdp:privs=proc_clock_highres

{global} newman@lenovo:~ $ getent exec_attr | grep -w xfreerdp
Basic Solaris User:solaris:cmd:::/usr/bin/xfreerdp:privs=proc_clock_highres

{global} newman@lenovo:~ $ ppriv -e -D /usr/bin/xfreerdp /size:800x600 /v:localhost:5001
[19:48:35:957] [21487:00000002] [INFO][com.freerdp.client.common.cmdline] - loading channelEx rdpdr
[19:48:35:957] [21487:00000002] [INFO][com.freerdp.client.common.cmdline] - loading channelEx rdpsnd
[19:48:35:958] [21487:00000002] [INFO][com.freerdp.client.common.cmdline] - loading channelEx cliprdr
[19:48:36:037] [21487:00000002] [ERROR][com.freerdp.core.nego] - Error: SSL_CERT_NOT_ON_SERVER
[19:48:36:040] [21487:00000002] [ERROR][com.freerdp.core.nego] - Error: SSL_CERT_NOT_ON_SERVER
[19:48:36:065] [21487:00000002] [INFO][com.freerdp.gdi] - Local framebuffer format  PIXEL_FORMAT_BGRX32
[19:48:36:065] [21487:00000002] [INFO][com.freerdp.gdi] - Remote framebuffer format PIXEL_FORMAT_RGB16
[19:48:36:229] [21487:00000002] [INFO][com.winpr.clipboard] - initialized POSIX local file subsystem
[19:48:36:278] [21487:00000008] [INFO][com.freerdp.channels.rdpsnd.client] - Loaded fake backend for rdpsnd
xfreerdp[21487]: missing privilege "proc_clock_highres" (euid = 101, syscall = 54) needed at secpolicy_clock_highres+0x1d
```

**Comment by AndWac** [comment]
You are dealing with profiles. I read that profiles are interpreted by a profile shell such as pfsh, pfksh and pfcsh. Have you tried to run the command in a profile shell or with pfexec?

**Comment by Mno-hime** [comment]
> Have you tried to run the command in a profile shell or with pfexec?

It works fine in `pfsh` and via `pfexec` even for a user with just the default execution profiles. So, I guess I create a `usr/bin/xfreerdp` script which runs `usr/bin/xfreerdp.bin` with `pfexec`, but I hoped this works more magically than that...

**Comment by Mno-hime** [comment]
> Have you tried to run the command in a profile shell or with pfexec?

It works fine in `pfsh` and via `pfexec` even for a user with just the default execution profiles. So, I guess I create a `usr/bin/xfreerdp` script which runs `usr/bin/xfreerdp.bin` with `pfexec`, but I hoped this works more magically than that...

**Comment by AndWac** [comment]
The magic should be there when you assign pfsh as his/her shell to the user.
Look at the time-slider package for an example.

**Comment by Mno-hime** [comment]
I think I can create `usr/bin/xfreerdp` script which does that under the hood and move the binary to `usr/bin/xfreerdp.bin`.

**Comment by 3eka** [comment]
I have tried to compile it (was able to do for 2.0.0-rc4),, but it has failed for final 2.0.0 version: https://github.com/FreeRDP/FreeRDP/issues/6105

Not sure how to proceed, since they have added mntent.h, which fails.
FYI

P.S: Testing OpenSSL 1.1.1 builds

**Comment by despair86** [comment]
yep ran into the same problem @3eka
(Oddly enough, freerdp is clean enough to compile with sun studio which is what cmake defaults to even on illumos)

**Pull request closed without merging**